runtime/caml/domain.h

```diff
@@ -113,6 +113,59 @@ int caml_try_run_on_all_domains(
   void*,
   void (*leader_setup)(caml_domain_state*));
 
+/* Function naming conventions for STW callbacks and STW critical sections.
+
+   A "STW callback" is a callback passed to one of the
+   [caml_try_run_on_all_domains*] runners, it will
+   run on all participant domains in parallel.
+
+   The "STW critical section" is the runtime interval betweeen the
+   start of the execution of the STW callback and the last barrier in
+   the callback. During this interval, mutator code from registered
+   participants cannot be running in parallel.
+
+   (Note that some parts of a STW callback are *not* inside the STW
+   critical section: all the code after the last barrier, or all the
+   callback if it does not contain a barrier.)
+
+   Some functions must be called within a STW critical section only,
+   calling then in a less-synchronized context introduces races with
+   mutators. To avoid these mistakes we use naming conventions as
+   a barebones effect system.
+
+   1. [stw_*] prefix for STW callbacks.
+
+   A function that defines a STW callback starts with [stw_] or
+   [caml_stw_]. It is passed to the [caml_try_run_on_all_domains*]
+   runner.
+
+   Examples:
+   - [caml_stw_empty_minor_heap] is a STW callback that empties the
+     minor heap
+   - [stw_resize_minor_heap_reservation] is a STW callback that
+     resizes the memory reservation for the minor heap
+
+   2. [*_from_stw] suffix for auxiliary functions that must be called
+      within a STW critical section.
+
+   3. [*_from_stw_single] suffix for auxiliary functions that must be
+      called by a *single* domain within a STW critical section --
+      typically the last one entering a barrier.
+
+      Note: [*_from_stw] and [*_from_stw_single] section may be safely
+      called during program initialization, when no mutators are
+      running and there are no domains running.
+
+   5. No [stw] in the name for functions that are not called in a STW
```

Comment:
Question: should we find a naming convention for functions starting a STW segment? Not obvious to me if the clarification is needed, but it does kind of sound more organic to me.

Reply:
My idea is that the naming convention helps when functions may only be called safely in certain contexts. For me, calling a STW pause is an implementation detail. For example a given runtime service may be implemented by either taking a lock on both reads and writes to an internal global resource, or by calling a STW pause on writes and doing nothing on reads. It makes sense to use the same function names for either implementations.

(For example in signals.c, `caml_install_signal_handler` uses a lock to protect a global variable; we might in the future want to update it to calling a STW section internally to avoid a race with `caml_execute_signal_exn`. But this needs not change how the callers refer to the function.)

Technically, calling a STW pause (directly or transitively) does restrict the contexts in which one can run, as STW pauses cannot be nested. This is in fact similar to taking locks, which also cannot be nested. But trying to enforce this discipline through naming would basically amount to marking half the runtime functions as "toplevel" functions and the other half as "leaf" functions, this would be invasive and not very useful.